modules/_labs/pagination/react/lib/Pagination/usePaginationModel.ts

```diff
@@ -0,0 +1,84 @@
+import * as React from 'react';
+
+import {buildPageRange} from './buildPageRange';
+import {PaginationModel} from './types';
+
+export type UsePaginationModelConfig = {
+  lastPage: number;
+  firstPage?: number;
+  initialCurrentPage?: number;
+  onPageChange?: (pageNumber: number) => void;
+  rangeSize?: number;
+};
+
+export const usePaginationModel = ({
+  firstPage = 1, // Q: Should this be a prop, or leave it out?
+  initialCurrentPage = 1,
+  lastPage,
+  rangeSize = 5,
+  onPageChange,
+}: UsePaginationModelConfig): PaginationModel => {
+  const [currentPage, setCurrentPage] = React.useState(initialCurrentPage);
+
+  React.useEffect(() => {
+    if (onPageChange) {
+      onPageChange(currentPage);
+    }
+  }, [onPageChange, currentPage]);
```

Comment:
`useEffect` causes `onPageChange` to be part of another render cycle. `useLayoutEffect` will cause `onPageChange` to be called as part of another state change batch. The compound component utility handles callbacks like this by defining a relationship between events and callbacks that are called in the same event queue vs another event queue.

You could mimic this here by making a function like the following and calling it instead of `setCurrentPage`:
```tsx
const changePage = (page: number) => {
  onPageChange?.(page);

  setCurrentPage(page)
}
```

An example use would look like:
```tsx
  const jumpToFirst = () => {
    changePage(firstPage);
  };
```

Reply:
The theme here is I don't like using `useEffect` for  state changes. It forces multiple renders of a component which is wasteful if it isn't necessary.

Reply:
Yeah, I agree. I think the reason I did this was because I was passing down `setCurrentPage` in the events, but maybe I don't need to do that. Or I could do something like:

```ts
const events = {
  setCurrentPage: changePage,
  jumpToFirst,
  jumpToLast,
  stepToNext,
  stepToPrevious,
  goToPage,
}
```

Reply:
I updated this. We're good now. 👍🏻